system/modules/playos-status.nix

```diff
@@ -0,0 +1,54 @@
+{ config, lib, pkgs, ... }:
+let
+  ttyNumber = 8;
+  tty = "tty${toString ttyNumber}";
+  ttyPath = "/dev/${tty}";
+  driverUrl = "http://127.0.0.1:8382";
+
+  script =
+    pkgs.writeShellScriptBin "print-status" ''
+      while :; do
+        printf "\033c"
+        screen=$(xrandr --current | grep '*' | awk '{print $1}')
+        network=$(connmanctl technologies | grep 'Type\|Connected')
```

Comment:
This prints information about all network interfaces. Should we filter for only WLAN?

Reply:
One context where this output will be used is to quickly check practical wifi availability when a Play Computer is combined with external antennae (typically when built into the Senso's tin console).

So it would be good to verify that wifi is not only available as a technology, but that actual wifi networks can be found. I inquired about the level of detail needed, but at a minimum I think we should count and print the number of available wireless networks. (Median) Signal strength might be of interest too, but I am not sure how effortful that is to obtain and whether it is worth it.

Reply:
Alright, I'll check what can be with reasonable effort.

Reply:
We can easily show the number of networks found. To calculate the median signal strength, I'd rewrite the script in Python, because it becomes too cumbersome to do in bash.

Another idea I had is that we could print the information to the tty from the controller, since it already has all the info related to networking.

Reply:
> Another idea I had is that we could print the information to the tty from the controller, since it already has all the info related to networking.

Interesting idea. How does that work, printing to a device file?

I guess one downside would be that getting the output from xrandr etc. would be a tad more cumbersome to do from OCaml?

Reply:
> Interesting idea. How does that work, printing to a device file?

I understood it differently, that it could be an HTTP request to the controller server.

Reply:
> Interesting idea. How does that work, printing to a device file?

Users in the `tty` group can simply write to the device file.

> I guess one downside would be that getting the output from xrandr etc. would be a tad more cumbersome to do from OCaml?

I've written very little OCaml, but it seems easy enough to do it with `Lwt_process`:
https://ocsigen.org/lwt/4.1.0/api/Lwt_process

> I understood it differently, that it could be an HTTP request to the controller server.

We could do that. The controller could expose an endpoint for information relevant to the status screen. The script could call that and merge it with data it collects from `xrandr` and the rest. Though I think maybe it would be simpler to do everything in one place.

Another idea I had is to let the controller serve an html page with the status info. It could then be viewed with a text-based browser like `w3m` on the tty. This is what `services.showManual` does in nixos.
An advantage of this would be that the content could be scrollable, if we'd want to show that much info (for example, a detailed list of all wifi networks). Another advantage would be that data collection could be done on page request (users would have to manually refresh), instead of at an interval, so it would be more resource efficient (not sure how much this matters in practice).

Reply:
Some quick thoughts:

- The HTML page + `w3m` approach seems attractive but [is a security problem](https://github.com/NixOS/nixpkgs/issues/83157). One would have to run a simple tool that is understood to be free of ways of executing commands or browsing the filesystem. `cat http://localhost:3333` comes to mind but isn't much different from the controller writing to the tty.
- I like the idea of writing from the controller to the tty. It does have the advantage of being able to rely on information that the controller gathers anyway.
- Anything that relies on the controller to control the content on the status tty has the downside that the tty becomes useless if the controller goes down. At the very least we would have to ensure that stale information is removed from the tty in that event.

All in all, I sense that while there may be potential to build something even more useful by relying on tools that can gather even more information, we can do something simple and already useful with the current approach, which seems dead simple and likely to be very robust.

Reply:
> The HTML page + w3m approach seems attractive but is a security problem. One would have to run a simple tool that is understood to be free of ways of executing commands or browsing the filesystem. cat http://localhost:3333 comes to mind but isn't much different from the controller writing to the tty.

Nice catch!

> Anything that relies on the controller to control the content on the status tty has the downside that the tty becomes useless if the controller goes down. At the very least we would have to ensure that stale information is removed from the tty in that event.

We could launch a one-off script with `systemd` to replace the contents of the tty with a warning if the controller goes down.

> All in all, I sense that while there may be potential to build something even more useful by relying on tools that can gather even more information, we can do something simple and already useful with the current approach, which seems dead simple and likely to be very robust.

I agree! For now I'll just refine the network info and later on we can look into re-implementing the functionality as part of the controller.